### PR TITLE
domain: record runaway and quarantine query

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #43691 

Problem Summary:

### What is changed and how it works?
Persist the runaway query records in KV data by the admin table

Once identified or watched quarantined are recorded in mysql.RUNAWAY_QUERIES. mysql.QUARANTINED_WATCH only active watch are kept
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
image
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -1246,6 +1248,121 @@ func (do *Domain) SetOnClose(onClose func()) {
...
+			// flush as soon as possible.
+			if len(quarantineRecordCh) == 0 || len(quarantineRecords) >= flushThrehold {
+				flushQuarantineRecords()
+			}
+		case r := <-recordCh:
+			records = append(records, r)
+			if len(records) >= flushThrehold {
+				flushRunawayRecords()
+			} else if fired {
+				// meet a new record, reset the timer.
```
Comment (Connor1996) [inline domain/domain.go:1303]: set fired = false here?

File: domain/resourcegroup/runaway.go [hunk 2]
```diff
@@ -104,6 +197,7 @@ func newRunawayChecker(manager *RunawayManager, resourceGroupName string, settin
 		deadline:          time.Now().Add(time.Duration(setting.Rule.ExecElapsedTimeMs) * time.Millisecond),
 		setting:           setting,
 		marked:            atomic.Bool{},
+		action:            strings.ToLower(setting.Action.String()),
```
Comment (Connor1996) [inline domain/resourcegroup/runaway.go:200]: why not get it from `setting` when needed

**Review by Connor1996 — COMMENTED** [review]
seems stale quarantined watch record is not cleaned

File: domain/resourcegroup/runaway.go (hunk 2 above)
Comment (CabinfeverB) [reply domain/resourcegroup/runaway.go:200]: I want to reduce the use of `strings.ToLower`.

**Comment by CabinfeverB** [comment]
> } else if fired {

Can we keep the history for now?  Users can use `select * from mysql.quarantined_watch where end_time >= now(); ` to query the active rules.

**Comment by CabinfeverB** [comment]
cc @glorv @nolouch

**Comment by Connor1996** [comment]
> > seems stale quarantined watch record is not cleaned
> 
> Can we keep the history for now? Users can use `select * from mysql.quarantined_watch where end_time >= now(); ` to query the active rules.

Is it temporary? otherwise I need to change the doc.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by CabinfeverB** [comment]
> 

IMO, it's not temporary. What do u think? @glorv

File: session/bootstrap.go [hunk 3]
```diff
@@ -604,6 +604,31 @@ const (
        KEY (create_time),
        KEY (create_user));`
 
+	// CreateRunawayTable stores the query which is identified as runaway or quarantined because of in watch list.
+	CreateRunawayTable = `CREATE TABLE IF NOT EXISTS mysql.runaway_queries (
```
Comment (nolouch) [inline session/bootstrap.go:608]:
```suggestion
	CreateRunawayTable = `CREATE TABLE IF NOT EXISTS mysql.tidb_runaway_queries (
```

I suggest using `tidb_xxx`  as the name Indicates tidb specific table.

File: session/bootstrap.go [hunk 4]
```diff
@@ -604,6 +604,31 @@ const (
...
+		action varchar(12) NOT NULL,
+		original_sql TEXT NOT NULL,
+		plan_digest TEXT NOT NULL,
+		tidb_server varchar(64),
+		INDEX plan_index(plan_digest(64)) COMMENT "accelerate the speed when add global binding query",
+		INDEX time_index(time) COMMENT "accelerate the speed when querying with active watch"
+	) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;`
+
+	// CreateQuarantineWatchTable stores the condition which is used to check whether query should be quarantined.
+	CreateQuarantineWatchTable = `CREATE TABLE IF NOT EXISTS mysql.quarantined_watch (
```
Comment (nolouch) [inline session/bootstrap.go:621]: How about `tidb_runaway_quarantined_watch `

**Comment by glorv** [comment]
> > 
> 
> IMO, it's not temporary. What do u think? @glorv

As each quarantined query has a TLL, I think that means they should be GCed over time. We may treat them as logs, gc them after a period of time, e.g. after 30 days or so.

File: session/bootstrap.go (hunk 3 above)
Comment (glorv) [reply session/bootstrap.go:608]: +1.

**Review by nolouch — APPROVED** [review]

**Review by glorv — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**